### PR TITLE
CORE-5324 CPI packaging tool should validate the CPB is valid

### PR DESCRIPTION
**Pull request by kyriathar**
Verifies it is a Cpb before packing it into a Cpi.

- adds Cpb verification checks to both `CreateCpi` (V1) and `CreateCpiV2` cli tools.
- enhances testing keystore to include certificates needed for Cpb signatures verification.
- replaces custom Cpb creation/ signing functions in testing with `TestCpbV1Builder` and `TestCpbV2Builder` apis.

### CODE REVIEW

File: tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpi.kt [hunk 1]
```diff
@@ -71,26 +76,45 @@ class CreateCpi : Runnable {
...
+            .format(VerifierFactory.FORMAT_1)
+            .name(cpbPath.toString())
+            .inputStream(FileInputStream(cpbPath.toString()))
+            .trustedCerts(readCertificates(signingOptions.keyStoreFileName, signingOptions.keyStorePass))
+            .build()
+            .verify()
+        true
+    } catch (e: Exception) {
+        System.err.println("Error verifying cpb: ${e.message}")
+        false
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpi.kt]: The program should exit with a non-zero exit code on error. I imagine throwing the exception would do that but also output a stack trace when we want this simple error message. A call to exit(1) after the println might be the best choice.

Calling exit here would save us from having the if statement wrapping all the code too. Hopefully the diff reduces down to just the call to this method.

Comment (kyriathar) [reply tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpi.kt]: Good point, will address.

File: tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/VerificationHelpers.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package net.corda.cli.plugins.packaging.signing
+
+import java.io.File
+import java.security.KeyStore
+import java.security.cert.X509Certificate
+import net.corda.cli.plugins.packaging.FileHelpers
+
+object VerificationHelpers {
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/VerificationHelpers.kt]: Names like `VerificationHelpers` and `SigningHelpers` are open ended. There is a lot that could go in there. It becomes inviting when a future developer is looking for a place to put new code. Over time, classes with names like these build up unrelated code.

If we give this a more specific name, we will reduce the chances of this accumulating new code.

As it stands, this class is reading certificates from a java keystore. So we could call it `JavaKeyStore`, `KeyStoreReader`, `CertificateReader` or `CertificateLoader`. It doesn't have to be one of those, feel free to think up a different name.

The aim is to capture the one (and only one) thing that this class does.

File: tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt [hunk 3]
```diff
@@ -15,6 +17,25 @@ import org.junit.jupiter.api.Assertions.assertNotNull
...
+        return this::class.java.getResourceAsStream(filePath)
+            ?: throw IllegalStateException("$filePath not found!")
+    }
+
+    val SIGNING_KEY_1_ALIAS = "signing key 1"
+    val SIGNING_KEY_2_ALIAS = "signing key 2"
+
+    val SIGNING_KEY_1 = privateKeyEntry(SIGNING_KEY_1_ALIAS, resourceInputStream("/signingkeys.pfx"))
+    val SIGNING_KEY_2 = privateKeyEntry(SIGNING_KEY_2_ALIAS, resourceInputStream("/signingkeys.pfx"))
+
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt]: This is an example of how names attract new code. 

These additions represent signing keys, we could move them to new class called `SigningKeys` or `TestSigningKeys`.

File: tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpi.kt (hunk 1 above)
Comment (kyriathar) [reply tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpi.kt]: Done -let the exception reach picocli code to be handled by the framework-.

File: tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt [hunk 4]
```diff
@@ -127,20 +88,15 @@ class CreateCpiTest {
...
                 "-p=keystore password",
-                "-k=signing key 1",
-                "--sig-file=$SIGNER_NAME"
+                "-k=$SIGNING_KEY_2_ALIAS",
+                "--sig-file=$CPI_SIGNER_NAME"
             )
-        }
-
-        // Expect output to be silent on success
-        assertEquals("", errText)
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt]: Why was this removed?

File: tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt [hunk 5]
```diff
@@ -105,20 +71,15 @@ class CreateCpiTest {
...
-                "--file=$outputFile",
-                "--keystore=${testKeyStore}",
-                "--storepass=keystore password",
-                "--key=signing key 1",
-                "--sig-file=$SIGNER_NAME"
-            )
-        }
-
-        // Expect output to be silent on success
-        assertEquals("", errText)
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt]: Another one here

File: tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/VerificationHelpers.kt (hunk 2 above)
Comment (kyriathar) [reply tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/VerificationHelpers.kt]: Renamed to `CertificateLoader`.

File: tools/plugins/package/src/test/resources/generate-keystore.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+# Root CA
+keytool -genkeypair -alias rootca -dname 'CN=Corda Dev Root CA, OU=R3, O=Corda, L=Dublin, C=IE' \
+        -keyalg ${KEY_ALG} -validity 4000 -keystore signingkeys.pfx -storetype pkcs12 -storepass "${STOREPASS}" \
+        -ext BasicConstraints:critical -ext KeyUsage=cRLSign,digitalSignature,keyCertSign
+
+keytool -genkeypair -alias "${SIGNING_KEY_1}" -dname "cn=CPI Plugin Example - Signing Key 1, o=R3, c=GB" \
+        -keyalg ${KEY_ALG} -validity 4000 -keystore signingkeys.pfx -storetype pkcs12 -storepass "${STOREPASS}"
+
+keytool -genkeypair -alias "${SIGNING_KEY_2}" -dname "cn=CPI Plugin Example - Signing Key 2, o=R3, c=GB" \
+        -keyalg ${KEY_ALG} -validity 4000 -keystore signingkeys.pfx -storetype pkcs12 -storepass "${STOREPASS}" \
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/test/resources/generate-keystore.sh:17]: We only need these two keys. We don't need to build a chain for this testing.

File: tools/plugins/package/src/test/resources/generate-keystore.sh [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+# Import signing key 2 certificate signed by root CA
+keytool -certreq -alias "${SIGNING_KEY_2}" -keystore signingkeys.pfx -storepass "${STOREPASS}" \
+        | keytool -gencert -alias rootca -rfc -keystore signingkeys.pfx -storepass "${STOREPASS}" \
+                  -ext BasicConstraints:critical -ext KeyUsage=cRLSign,digitalSignature,keyCertSign > key2.pem
+cat rootca.pem key2.pem > key2-chain.pem
+keytool -importcert -alias "${SIGNING_KEY_2}" -file key2-chain.pem -noprompt -keystore signingkeys.pfx -storepass "${STOREPASS}"
+
+# To trust freetsa.org, download their CA cert and import into keystore
+keytool -importcert -alias freetsa -keystore signingkeys.pfx -storepass "${STOREPASS}" -file ~/Downloads/cacert.pem
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/test/resources/generate-keystore.sh]: We should remove this. It is not needed for the tests, relies on ~/Downloads and trusts an external service.

File: tools/plugins/package/src/test/resources/generate-keystore.sh [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+rm *.pem
```
Comment (josephzunigadaly) [inline tools/plugins/package/src/test/resources/generate-keystore.sh]: This assumes all `.pem` files were generated by this script. It could be dangerous. We should list specific pem files we know the script generates or remove the line entirely.

File: tools/plugins/package/src/test/resources/generate-keystore.sh (hunk 6 above)
Comment (josephzunigadaly) [reply tools/plugins/package/src/test/resources/generate-keystore.sh:17]: Actually it is closer to real usage with the chain. Forget I said this. No need to change anything for this one.

File: tools/plugins/package/src/test/resources/generate-keystore.sh (hunk 8 above)
Comment (kyriathar) [reply tools/plugins/package/src/test/resources/generate-keystore.sh]: Done

File: tools/plugins/package/src/test/resources/generate-keystore.sh (hunk 7 above)
Comment (kyriathar) [reply tools/plugins/package/src/test/resources/generate-keystore.sh]: Good catch, done.

File: tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt (hunk 3 above)
Comment (kyriathar) [reply tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt]: Done

File: tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt (hunk 4 above)
Comment (kyriathar) [reply tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt]: Putting those back

**Review by josephzunigadaly — DISMISSED** [review]

**Review by josephzunigadaly — APPROVED** [review]
